File: day1/day1.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <cstdlib>

struct Instruction {
    char dir; // 'L' or 'R'
    int steps;
};

int wrap(int x) {
    x %= 100;
    if (x < 0) x += 100;
    return x;
}

Instruction parse(const std::string& s) {
    Instruction ins;
    ins.dir = s[0];
    ins.steps = std::atoi(s.c_str() + 1);
    return ins;
}

std::vector<std::string> read_input(const char* filename) {
    std::ifstream f(filename);
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(f, line)) {
        if (!line.empty())
            lines.push_back(line);
    }
    return lines;
}

// Counts zero crossings if count_all is true, else counts landing on zero
long solve(const std::vector<std::string>& lines, bool count_all) {
    std::vector<Instruction> instrs;
    for (size_t i = 0; i < lines.size(); ++i)
        instrs.push_back(parse(lines[i]));

    int pos = 50;
    long pw = 0;

    for (size_t i = 0; i < instrs.size(); ++i) {
        Instruction ins = instrs[i];
        int delta = (ins.dir == 'L') ? 1 : -1;

        if (count_all) {
            for (int step = 1; step <= ins.steps; ++step) {
                int p = wrap(pos + delta * step);
                if (p == 0) ++pw;
            }
        }

        pos = wrap(pos + (ins.dir == 'L' ? ins.steps : -ins.steps));

        if (!count_all && pos == 0)
            ++pw;
    }

    return pw;
}

int main() {
    std::vector<std::string> lines = read_input("input.txt");

    long part1 = solve(lines, false);
    long part2 = solve(lines, true);

    std::cout << "Part 1: " << part1 << ", Part 2: " << part2 << "\n";
}

File: day1/day1.py
"""
Day 1, Advent of Code
"""

from pathlib import Path


def compute_dial_position(position, direction, steps, total_positions=100):
    """Computes the new position on a combination dial."""
    delta = steps if direction == "L" else -steps
    return (position + delta) % total_positions


def count_zero_crossings(start, steps, direction, total_positions=100):
    """Count how many times we click through zero."""
    delta = 1 if direction == "L" else -1
    return sum((start + delta * i) % total_positions == 0 for i in range(1, steps + 1))


def solve_combination_lock(instructions, count_crossings=False, total_positions=100):
    """Solve the combination lock puzzle.

    Args:
        instructions: List of direction+steps strings (e.g., ['L68', 'R30'])
        count_crossings: If True, count all passes through zero.
                        If False, only count when landing on zero.
    """
    position = 50
    password = 0

    for instruction in instructions:
        direction, steps = instruction[0], int(instruction[1:])

        if count_crossings:
            password += count_zero_crossings(
                position, steps, direction, total_positions
            )

        position = compute_dial_position(position, direction, steps, total_positions)

        if not count_crossings and position == 0:
            password += 1

    return password


def read_input(filename: Path) -> str:
    """Given a set of combination instructions, read the input file"""
    with open(filename) as f:
        lines = [line.strip() for line in f.readlines()]
    return lines


if __name__ == "__main__":
    # pt1 = solve_combination_lock(read_input(Path('test_input.txt')), count_crossings=False)
    # pt2 = solve_combination_lock(read_input(Path('test_input.txt')), count_crossings=True)
    # print(f'Part 1: {pt1}, Part 2: {pt2}')
    pt1 = solve_combination_lock(read_input(Path("input.txt")), count_crossings=False)
    pt2 = solve_combination_lock(read_input(Path("input.txt")), count_crossings=True)
    print(f"Part 1: {pt1}, Part 2: {pt2}")
